Various fixes to broken optional patches

auburnbit opened the pull request:
Please see commits for details.
(fixes #11, fixes #12)

auburnbit committed:
Updated MMC1 OriginalGFX.asm for 3.0

Updated the starting address of the animated tile frames which had been shifted a few bytes between 2.0 and 3.0. This was causing a crash upon loading if this patch was used.

Added code to return Ganon's palette after he recovers from his vulnerable state to the original palette, as 3.0 modified this and it hadn't been accounted for here.

Updated "Original NES GFX.ips".

auburnbit committed:
Updated MMC1 LinksAwakeningGFX optional patch for 3.0

Updated the starting address of the animated tile frames which had been shifted a few bytes between 2.0 and 3.0. This was causing graphical glitches if this patch was used.

Updated corresponding IPS patch with changes.

auburnbit committed:
Fixed crash when using MMC5 OriginalHUD optional patch

Narrowed down problem code in OriginalHUD.asm

Two lines:
org $9E3D	// 0x19E4D
	cmp.b #$F6	// Finishing PPU low byte
were causing the crash when entering the cave.

Changed to:
org $9E3D	// 0x19E4D
	cmp.b #$E2	// Finishing PPU low byte

This is because move_maps.asm used  #$76 + #$80 to calculate #$F6 but this didn't get updated in OriginalHUD.asm where the new starting PPU x-position of the automap is #$62 instead.

ShadowOne333 commented:
Sorry for the very late reply, I did see your issue reports but haven't been able to respond to them yet due to vacations.
So far it all seems good, the only change I have doubts about is the one inside the Original HUD code for MMC5 ([MMC5/code/optional/OriginalHUD.asm](https://github.com/ShadowOne333/The-Legend-of-Zelda-Redux/pull/13/files#diff-f329137fc9aec40cf79eb77adae43da52e3fea2e57c45c20a83e1ae0997a3fee))

Have you verified and confirmed that commenting out those PPU related opcodes doesn't have any undesired drawbacks in-game?
Specifically these ones:
Original code from Move maps (MMC5):
https://github.com/ShadowOne333/The-Legend-of-Zelda-Redux/blob/master/MMC5/code/gameplay/move_maps.asm#L36-L49

Your change:
```
//org $9DC5	// 0x19DD5
//	adc.b #$62	// Starting PPU low byte
//org $9E22	// 0x19E32
//	ldy.b #$62	// Starting PPU low byte
//org $9E3D	// 0x19E4D
//	cmp.b #$F6	// Finishing PPU low byte
```

The one change at 9E3D doesn't seem to matter (since it's the same in both instances), but the other 2 have me in doubt as I believe these ones are the ones that set the starting X position for the Automap for MMC5. The original game used 2062 in PPU, while Bogaa's MMC5 implementation of Automap used 2076.

Once this one is confirmed, the rest can be easily be merged.
Thanks for taking a look at this!

auburnbit commented:
Agreed on the uncertainty on the Original HUD code fix. I can say that it does stop the crash and allows the game to play, so I guess it's better than the alternative. However, I haven't done extensive testing or paid close attention to whether everything about the Automap position is correct. I'll play around with it some more and see what I can figure out.

auburnbit commented:
It looks like the one change at 9E3D was the problem, commenting that out alone seems to stop the crash. I'm curious why modifying 9E3D makes it crash when Move Maps seems to make the same change without issue.

auburnbit commented:
Wait... I bet $9E3D's supposed to be 
`cmp.b #$E2`
It's working as it is with that line commented out but I'll test it with the above change instead.

Edit: Yep that was it. I've updated the commit with the details. 

Edit2: Also double checked the three new IPS files to be sure I didn't mess something up and they seem to be correct. Also, it appears that there actually is no need to comment move_maps.asm out of main.asm when using OriginalHUD.asm (as it is now at least), as it overwrites everything move_maps.asm does (in both MMC1 and MMC5). As long as that remains true in future iterations of both files, there shouldn't be a need to comment out move_maps.asm when using Original.HUD.asm. I've removed the warnings I had added to the optional.asm files and updated the commit.

(I apologize for all the edits and force pushes, this is my first time actually using GitHub so I'm still figuring things out.)

auburnbit commented on MMC5/code/optional/OriginalHUD.asm:
```diff
@@ -86,7 +86,7 @@ org $9DC5	// 0x19DD5
 org $9E22	// 0x19E32
 	ldy.b #$62	// Starting PPU low byte
 org $9E3D	// 0x19E4D
-	cmp.b #$F6	// Finishing PPU low byte
```
#$F6 comes from adding #$80 to #$76 (the starting ppu low byte in move_maps.asm) and wasn't updated when the starting ppu low byte was changed to #$62 in this file, causing a crash when entering the cave

auburnbit commented on MMC5/code/optional/OriginalHUD.asm:
```diff
@@ -86,7 +86,7 @@ org $9DC5	// 0x19DD5
 	ldy.b #$62	// Starting PPU low byte
 org $9E3D	// 0x19E4D
-	cmp.b #$F6	// Finishing PPU low byte
+	cmp.b #$E2	// Finishing PPU low byte
```
adding #$80 to #$62 gives us the correct finishing PPU low byte, #$E2, and fixes the crash

ShadowOne333 commented:
Gave these a look with the new modifications, and it all seems good!
Thank you so much for narrowing down the issues and debugging them to find a proper fix for these.
I'll merge this and update the optional patches in Romhacking.net as well with your fixes, I'll also add you in the credits for the fixes.

Just to be sure, the only IPS files modified with Link's Awakening GFX, Original GFX (MMC1) and Original HUD (MMC5), right?

auburnbit commented:
Yes, that should be the only IPS files modified. Thanks!
